rfcs/0000-local-overlay-store.md

```diff
@@ -0,0 +1,241 @@
+---
+feature: local-overlay-store
+start-date: (fill me in with today's date, YYYY-MM-DD)
+author: (name of the main author)
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Add a new `local-overlay` store implementation to Nix.
+This will be a local store that is layered upon another local filesystem store (local store or daemon).
+This allows locally extending a shared store that is periodically extended with additional store objects.
+
+# Motivation
+[motivation]: #motivation
+
+TODO: link Replit blog post.
+
+## Technical motivation
+
+Many organizational users of Nix have a large collection of Nix store objects they wish to share with a number of consumers, be they human users, build farm workers etc.
+The existing ways of doing this are:
+
+- Share nothing: Copy store objects to each consumer
+
+- Share everything: Use single mutable NFS store, carefully synchronizing updates.
+
+- Overlay everything: Mount all of `/nix` (store and DB) with OverlayFS.
+
+Each has serious drawbacks:
+
+- Share nothing wastes tones of space as many duplicate store objects are stored separately.
+
+- Share everything incurs major overhead from synchronization, even if consumers are making store objects they don't intend any other consumer to use.
+
+- Overlay everything cannot take advantage of new store objects added to the lower store, because its "fork" of the DB covers up the lower store's.
+
+The new `local-overlay` store also uses OverlayFS, but just for the store directory.
+The database is still a regular fresh empty one to start, and instead Nix explicitly knows about the lower store, so it can get any information for the DB it needs from it manually.
+This avoids all 3 downsides:
+
+- Store objects are never deduplicated.
+  OverlayFS ensures that one copy in either store dir layer is enough, and we are careful never to wastefully include a store object in both layers.
+
+- No excess synchronization.
+  local changes are just that: local, not shared with any other consumer.
+  The lower store is never written to (no modifications or even filesystem locks) so any slow NFS write and sync paths should not be encountered.
```

Comment:
> so any slow NFS write and sync paths should not be encountered.

I'm having a little trouble parsing this sentence properly. Maybe reorder it to be clearer; something like:

> so we should not encounter any slow NFS writes or path syncs.